# Ex 4.9 Intro to Visulization with Python Part 1

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path=r'C:\Users\jecka\2023-09-11 Achievement 4 Python'

In [3]:
#Import customer dataframe
df_customer = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

## Data Wrangling

In [4]:
df_customer.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [5]:
df_customer.shape

(206209, 10)

In [6]:
df_customer.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [7]:
df_customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [8]:
df_customer.value_counts(dropna = False)

user_id  First Name  Surnam    Gender  STATE          Age  date_joined  n_dependants  fam_status                        income
1        Linda       Nguyen    Female  Alabama        31   2/17/2019    3             married                           40423     1
137478   Elizabeth   Mcdonald  Female  Maryland       19   12/16/2019   2             living with parents and siblings  28625     1
137468   Todd        Welch     Male    Nevada         31   7/11/2017    1             married                           42533     1
137469   Douglas     Andrews   Male    New Hampshire  45   4/18/2018    1             married                           125718    1
137470   Jeremy      Bennett   Male    New Jersey     19   5/10/2019    3             living with parents and siblings  79969     1
                                                                                                                                 ..
68740    Michael     Wall      Male    Tennessee      48   8/18/2017    3        

In [9]:
df_customer['Gender'].value_counts(dropna = False)

Gender
Male      104067
Female    102142
Name: count, dtype: int64

In [10]:
df_customer['date_joined'].value_counts(dropna = False)

date_joined
9/17/2018     213
2/10/2018     212
4/1/2019      211
9/21/2019     211
12/19/2017    210
             ... 
9/1/2018      141
1/22/2018     140
11/24/2017    139
7/18/2019     138
8/6/2018      128
Name: count, Length: 1187, dtype: int64

In [11]:
df_customer['Age'].value_counts(dropna = False)

Age
19    3329
55    3317
51    3317
56    3306
32    3305
      ... 
65    3145
25    3127
66    3114
50    3102
36    3101
Name: count, Length: 64, dtype: int64

In [12]:
df_customer['First Name'].value_counts(dropna = False)

First Name
NaN        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: count, Length: 208, dtype: int64

In [13]:
df_customer['Surnam'].value_counts(dropna = False)

Surnam
Hamilton      252
Randall       248
Lamb          243
Pennington    243
Barnett       242
             ... 
Poole         172
Bauer         166
Pearson       164
Payne         163
Jordan        162
Name: count, Length: 1000, dtype: int64

#### Insights:
##### Surname is mistyped, so that needs to be changed. I may just change both name columns to first and last
##### just to keep it uniformed I'll change the columns with capital letters (ex STATE to "state")
##### The ages range from 18 to 81
##### User_id can be linked with the merged Orders/Products df
##### There are 11259 first names left blank

In [14]:
# Renaming Columns
df_customer.rename(columns = {'First Name' : 'first'}, inplace = True)

In [15]:
df_customer.rename(columns = {'Surnam' : 'last'}, inplace = True)

In [16]:
df_customer.rename(columns = {'Gender' : 'gender'}, inplace = True)

In [17]:
df_customer.rename(columns = {'STATE' : 'state'}, inplace = True)

In [18]:
df_customer.rename(columns = {'Age' : 'age'}, inplace = True)

In [19]:
df_customer.rename(columns = {'n_dependants' : 'number_of_dependents'}, inplace = True)

In [20]:
df_customer.rename(columns = {'fam_status' : 'marital_status'}, inplace = True)

In [21]:
df_customer.head()

user_id     first      last  gender       state  age date_joined  \
0    26711   Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890  Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803   Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935  Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797       Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependents marital_status  income  
0                     3        married  165665  
1                     0         single   59285  
2                     2        married   99568  
3                     0         single   42049  
4                     1        married   40374

In [22]:
#Checking for duplicates
df_dup = df_customer[df_customer.duplicated()]

In [23]:
df_dup

Empty DataFrame
Columns: [user_id, first, last, gender, state, age, date_joined, number_of_dependents, marital_status, income]
Index: []

There are no duplicates!!

In [24]:
df_customer.loc[df_customer['first'].isnull() == True]

user_id first      last  gender         state  age date_joined  \
53        76659   NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738   NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996   NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166   NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778   NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...   ...       ...     ...           ...  ...         ...   
206038   121317   NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799   NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394   NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532   NaN     Floyd  Female    California   39    4/1/2020   
206171   116898   NaN   Delgado  Female      Colorado   23    4/1/2020   

        number_of_dependents marital_status  income  
53                         2        married   41709  
73                         0         single   82518  
82                         3        married  117099  
99                         1        married  155673  
105                        3        married  151819  
...                      ...            ...     ...  
206038                     3        married   87783  
206044                     2        married  108488  
206090                     1        married   45275  
206162                     0         single   56325  
206171                     2        married   59222  

[11259 rows x 10 columns]

In [25]:
# Checking if there are mixed value types
for col in df_customer.columns.tolist():
  weird = (df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customer[weird]) > 0:
    print (col)

first


In [26]:
df_customer['first'].fillna('Unknown', inplace=True)

In [27]:
# Checking once more if there are mixed value types
for col in df_customer.columns.tolist():
  weird = (df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customer[weird]) > 0:
    print (col)

In [28]:
df_customer['first'].value_counts()

first
Unknown    11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: count, Length: 208, dtype: int64

In [29]:
# final check for null
df_customer.isnull().sum()

user_id                 0
first                   0
last                    0
gender                  0
state                   0
age                     0
date_joined             0
number_of_dependents    0
marital_status          0
income                  0
dtype: int64

## Combining with Orders/Products df

In [30]:
# import merged dataset
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_group_transform.pkl'))

In [31]:
# Checking id user_id are same data type
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 25 columns):
 #   Column                 Dtype   
---  ------                 -----   
 0   order_id               int64   
 1   user_id                int64   
 2   eval_set               object  
 3   order_number           int64   
 4   orders_day_of_week     int64   
 5   order_hour_of_day      int64   
 6   days_since_last_order  float64 
 7   product_id             float64 
 8   add_to_cart_order      float64 
 9   reordered              float64 
 10  product_name           object  
 11  aisle_id               int64   
 12  department_id          int64   
 13  prices                 float64 
 14  _merge                 category
 15  price_range_loc        object  
 16  busiest_day            object  
 17  busiest_days           object  
 18  busiest_period_of_day  object  
 19  max_order              int64   
 20  loyalty_flag           object  
 21  avg_price              float6

In [32]:
df_customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  int64 
 1   first                 206209 non-null  object
 2   last                  206209 non-null  object
 3   gender                206209 non-null  object
 4   state                 206209 non-null  object
 5   age                   206209 non-null  int64 
 6   date_joined           206209 non-null  object
 7   number_of_dependents  206209 non-null  int64 
 8   marital_status        206209 non-null  object
 9   income                206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [ ]:
customer_ords_prods = df_customer.merge(ords_prods_merge, on = ['user_id'])

In [36]:
customer_ords_prods.shape

(32404859, 34)

In [37]:
customer_ords_prods.head()

user_id    first      last  gender     state  age date_joined  \
0    26711  Deborah  Esquivel  Female  Missouri   48    1/1/2017   
1    26711  Deborah  Esquivel  Female  Missouri   48    1/1/2017   
2    26711  Deborah  Esquivel  Female  Missouri   48    1/1/2017   
3    26711  Deborah  Esquivel  Female  Missouri   48    1/1/2017   
4    26711  Deborah  Esquivel  Female  Missouri   48    1/1/2017   

   number_of_dependents marital_status  income  ...    price_range_loc  \
0                     3        married  165665  ...  Mid-range product   
1                     3        married  165665  ...  Mid-range product   
2                     3        married  165665  ...  Mid-range product   
3                     3        married  165665  ...  Mid-range product   
4                     3        married  165665  ...  Mid-range product   

      busiest_day    busiest_days  busiest_period_of_day  max_order  \
0  Regularly busy     Busiest day            Most orders          8   
1  Regularly busy  Regularly busy            Most orders          8   
2  Regularly busy     Busiest day            Most orders          8   
3  Regularly busy  Regularly busy            Most orders          8   
4  Regularly busy      Least busy            Most orders          8   

   loyalty_flag  avg_price  spender_type  past_orders_median    frequency_flag  
0  New customer   7.988889   Low spender                19.0  Regular customer  
1  New customer   7.988889   Low spender                19.0  Regular customer  
2  New customer   7.988889   Low spender                19.0  Regular customer  
3  New customer   7.988889   Low spender                19.0  Regular customer  
4  New customer   7.988889   Low spender                19.0  Regular customer  

[5 rows x 34 columns]

In [39]:
# Export to pkl
customer_ords_prods.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'customer_ords_prods_merge.pkl'))